Support organisation runners & add more variables.

raj-saxena opened the pull request:
@maintainers, thanks for the base work!

We needed Organisation Runners and followed the advice in this discussion (https://github.com/terraform-google-modules/terraform-google-github-actions-runners/issues/3) to make the necessary changes.
Had to make `repo_url` & `repo_name` optional to support Organisation Runners. If `repo_name` is empty, organisation runners are registered.

We also added the ability to customize few other parameters like:
1. Ability to provide `network` but still let the module create `sub-network`
2. Ability to pass `region` instead of hardcoding it to `US`. Happy to change this to a list and then add `dynamic` block if needed.
3. Ability to pass `labels` to the module that the runners can be tagged with during registration.
4. Bug - The `target_size` should not be specified as autoscaling is enabled. From the [managed-instance-group](https://github.com/terraform-google-modules/terraform-google-vm/tree/v7.1.0/modules/mig) docs -
>  target_size | The target number of running instances for this managed instance group. This value should always be explicitly set unless this resource is attached to an autoscaler, in which case it should never be set.
5. Instead of above, added support for `min_replicas` & `max_replicas`.
6. Updated `source_image_project` from the deprecated `gce-uefi-images` family.
7. Updated to the latest version of github action `2.283.2`

If the changes make sense, it would be nice to get them merged into the module. 
Happy to receive any feedbacks.

comment-bot-dev commented:
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/gh-runner-mig-vm/README.md /tmp/tmp.9VUmtR46U2/generate_docs/workspace/modules/gh-runner-mig-vm/README.md
33a34
> | create\_subnetwork | Whether to create subnetwork or use the one provided via subnet\_name | `bool` | `true` | no |
36a38
> | labels | Labels to attach to the runners | `set(string)` | n/a | yes |
37a40,41
> | max\_replicas | Maximum number of runner instances | `number` | `10` | no |
> | min\_replicas | Minimum number of runner instances | `number` | `2` | no |
41c45
< | repo\_name | Name of the repo for the Github Action | `string` | n/a | yes |
---
> | repo\_name | Name of the repo for the Github Action | `string` | `""` | no |
43c47
< | repo\_url | Repo URL for the Github Action | `string` | n/a | yes |
---
> | repo\_url | Repo URL for the Github Action | `string` | `""` | no |
49c53
< | source\_image\_project | Project where the source image comes from | `string` | `"gce-uefi-images"` | no |
---
> | source\_image\_project | Project where the source image comes from | `string` | `"ubuntu-os-cloud"` | no |
54d57
< | target\_size | The number of runner instances | `number` | `2` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_shell
 Failed shell check. More info on running shellcheck locally [here](https://www.shellcheck.net). 
 ```bash 
Running shellcheck
In ./modules/gh-runner-mig-vm/scripts/startup.sh line 50:
    RUNNER_ALLOW_RUNASROOT=1 /runner/config.sh --unattended --url https://github.com/${REPO_OWNER} --replace --work "/runner-tmp" --token "$ACTIONS_RUNNER_INPUT_TOKEN" --labels $LABELS
                                                                                     ^-----------^ SC2086: Double quote to prevent globbing and word splitting.
                                                                                                                                                                                 ^-----^ SC2086: Double quote to prevent globbing and word splitting.
In ./modules/gh-runner-mig-vm/scripts/startup.sh line 55:
    RUNNER_ALLOW_RUNASROOT=1 /runner/config.sh --unattended --url https://github.com/${REPO_OWNER}/${REPO_NAME} --replace --work "/runner-tmp" --url "$REPO_URL" --token "$ACTIONS_RUNNER_INPUT_TOKEN" --labels $LABELS
                                                                                     ^-----------^ SC2086: Double quote to prevent globbing and word splitting.
                                                                                                   ^----------^ SC2086: Double quote to prevent globbing and word splitting.
                                                                                                                                                                                                                ^-----^ SC2086: Double quote to prevent globbing and word splitting.
For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/gh-runner-mig-vm/main.tf
--- old/modules/gh-runner-mig-vm/main.tf
+++ new/modules/gh-runner-mig-vm/main.tf
@@ -15,11 +15,11 @@
  */
 
 locals {
-  network_name       = var.create_network ? google_compute_network.gh-network[0].self_link : var.network_name
-  subnet_name        = var.create_subnetwork ? google_compute_subnetwork.gh-subnetwork[0].self_link : var.subnet_name
-  service_account    = var.service_account == "" ? google_service_account.runner_service_account[0].email : var.service_account
-  startup_script     = var.startup_script == "" ? file("${path.module}/scripts/startup.sh") : var.startup_script
-  shutdown_script    = var.shutdown_script == "" ? file("${path.module}/scripts/shutdown.sh") : var.shutdown_script
+  network_name    = var.create_network ? google_compute_network.gh-network[0].self_link : var.network_name
+  subnet_name     = var.create_subnetwork ? google_compute_subnetwork.gh-subnetwork[0].self_link : var.subnet_name
+  service_account = var.service_account == "" ? google_service_account.runner_service_account[0].email : var.service_account
+  startup_script  = var.startup_script == "" ? file("${path.module}/scripts/startup.sh") : var.startup_script
+  shutdown_script = var.shutdown_script == "" ? file("${path.module}/scripts/shutdown.sh") : var.shutdown_script
 }
 
 /*****************************************
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

raj-saxena commented:
Not sure how to resolve the warnings above.

bharathkkb commented on modules/gh-runner-mig-vm/variables.tf:
```diff
@@ -68,19 +75,38 @@ variable "restart_policy" {
 variable "repo_url" {
   type        = string
   description = "Repo URL for the Github Action"
+  default     = ""
```
Lets remove this var since we can use the approach above

bharathkkb commented on modules/gh-runner-mig-vm/variables.tf:
```diff
@@ -68,19 +75,38 @@ variable "restart_policy" {
   description = "Owner of the repo for the Github Action"
 }
 
+variable "labels" {
+  type        = set(string)
+  description = "Labels to attach to the runners"
+  default     = []
+}
```
```suggestion
variable "gh_runner_labels" {
  type        = set(string)
  description = "GitHub runner labels to attach to the runners. Docs: https://docs.github.com/en/actions/hosting-your-own-runners/using-labels-with-self-hosted-runners"
  default     = []
}
```

bharathkkb reviewed (changes requested):
Thanks for the PR, some minor changes

raj-saxena commented:
Thanks for the review @bharathkkb . I have addressed your comments in this commit https://github.com/terraform-google-modules/terraform-google-github-actions-runners/pull/34/commits/e4d8feb47407435f106613eb90cb744d3a52011b
